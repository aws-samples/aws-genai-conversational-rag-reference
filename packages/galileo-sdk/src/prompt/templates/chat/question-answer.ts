/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
import { BASE_CHAT_PARTIALS, ChatTemplatePartials } from './base.js';
import { HandlebarsPromptTemplate, HandlebarsPromptTemplateRuntime, ScopedHandlebarsPromptTemplateInput } from '../../handlebars.js';

export const CHAT_QUESTION_ANSWER_PARTIALS: ChatTemplatePartials = {
  ...BASE_CHAT_PARTIALS,
  Context: '{{>Corpus}}',
  Instruction: 'You are a research assistant in the "{{domain}}" domain. Based on the following rules and provided corpus denoted by {{>Delimiter}}, answer the question. \nRules:\n{{>Rules}}',
  Cue: 'Question: {{question}}\n\nAnswer: ',
} as const;


export interface ChatQuestionAnswerPromptTemplateInputValues {
  readonly domain: string;
  readonly context: string;
  readonly question: string;
  readonly rules?: string[];
}

export type ChatQuestionAnswerPromptTemplateInput =
  ScopedHandlebarsPromptTemplateInput<ChatTemplatePartials, ChatQuestionAnswerPromptTemplateInputValues>;
export type ChatQuestionAnswerPromptRuntime =
  HandlebarsPromptTemplateRuntime<ChatQuestionAnswerPromptTemplateInput>;

export class ChatQuestionAnswerPromptTemplate extends
  HandlebarsPromptTemplate<ChatTemplatePartials, ChatQuestionAnswerPromptTemplateInputValues> {

  static async deserialize(data: any) {
    return new ChatQuestionAnswerPromptTemplate(data);
  }

  constructor(input: ChatQuestionAnswerPromptTemplateInput) {
    super({
      template: '{{>Layout}}',
      inputVariables: ['context', 'domain', 'question'],
      ...input,
      partialVariables: {
        // LangChain partials only supports string or function return string... but actually can return anything
        rules: [
          'only use knowledge from the provided corpus to answer the question',
          'always be truthful, honest, unbiased, and unharmful',
          'be concise, do not repeat the question or yourself in the answer',
        ],
        ...input.partialVariables,
      },
      templatePartials: {
        ...CHAT_QUESTION_ANSWER_PARTIALS,
        ...input.templatePartials,
      },
    });
  }
}
