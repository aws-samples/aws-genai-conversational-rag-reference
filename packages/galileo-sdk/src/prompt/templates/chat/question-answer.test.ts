/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
PDX-License-Identifier: Apache-2.0 */
import { ChatQuestionAnswerPromptTemplate } from './question-answer.js';

describe('prompt/templates/chat', () => {
  describe('question-answer', () => {
    test('default', async () => {
      const template = new ChatQuestionAnswerPromptTemplate({});

      expect(
        await template.format({
          question: 'THE QUESTION',
          domain: 'DOMAIN',
          context: 'CORPUS',
        }),
      ).toMatchSnapshot();
    });

    test('messages', async () => {
      const template = new ChatQuestionAnswerPromptTemplate({
        template: `
{{#*inline "HumanMessage"}}<|prompter|>{{content}}<|endoftext|>{{/inline}}
{{#*inline "AIMessage"}}<|assistant|>{{content}}<|endoftext|>{{/inline}}
{{#*inline "SystemMessage"}}<|system|>{{content}}<|endoftext|>{{/inline}}

{{>Layout}}
        `,
      });

      expect(
        await template.format({
          question: 'THE QUESTION',
          domain: 'DOMAIN',
          context: 'CORPUS',
        }),
      ).toMatchSnapshot();
    });

    test('wrappers', async () => {
      const template = new ChatQuestionAnswerPromptTemplate({
        template: `
{{#*inline "Header"}}<SEQUENCE>{{/inline}}
{{#*inline "Footer"}}</SEQUENCE>{{/inline}}

{{>Layout}}
        `,
      });

      expect(
        await template.format({
          question: 'THE QUESTION',
          domain: 'DOMAIN',
          context: 'CORPUS',
        }),
      ).toMatchSnapshot();
    });

    test('should allow complete override of template', async () => {
      const template = new ChatQuestionAnswerPromptTemplate({
        template: 'CUSTOM: {{domain}} {{question}} {{context}}',
      });

      expect(
        await template.format({
          question: 'THE QUESTION',
          domain: 'DOMAIN',
          context: 'CORPUS',
        }),
      ).toBe('CUSTOM: DOMAIN THE QUESTION CORPUS');
    });

    test('custom rules', async () => {
      const template = new ChatQuestionAnswerPromptTemplate({});

      expect(
        await template.format({
          question: 'THE QUESTION',
          domain: 'DOMAIN',
          context: 'CORPUS',
          rules: ['rule 1', 'rule 2'],
        }),
      ).toMatchSnapshot();
    });

    test('partial template overrides', async () => {
      const template = new ChatQuestionAnswerPromptTemplate({
        templatePartials: {
          Header: '<WRAPPER>',
          Footer: '</WRAPPER>',
        },
      });

      expect(
        await template.format({
          question: 'THE QUESTION',
          domain: 'DOMAIN',
          context: 'CORPUS',
        }),
      ).toMatchSnapshot();
    });

    test('flatten', async () => {
      const template = new ChatQuestionAnswerPromptTemplate({});

      expect(template.flatten()).toMatchSnapshot();
    });
  });
});
