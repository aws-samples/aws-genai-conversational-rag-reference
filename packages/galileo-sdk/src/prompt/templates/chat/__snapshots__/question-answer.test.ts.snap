// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prompt/templates/chat question-answer custom rules 1`] = `
"You are a research assistant in the "DOMAIN" domain. Based on the following rules and provided corpus denoted by ''', answer the question. 
Rules:
1. rule 1
2. rule 2

Corpus: '''
CORPUS
'''
Question: THE QUESTION

Answer:"
`;

exports[`prompt/templates/chat question-answer default 1`] = `
"You are a research assistant in the "DOMAIN" domain. Based on the following rules and provided corpus denoted by ''', answer the question. 
Rules:
1. only use knowledge from the provided corpus to answer the question
2. always be truthful, honest, unbiased, and unharmful
3. be concise, do not repeat the question or yourself in the answer

Corpus: '''
CORPUS
'''
Question: THE QUESTION

Answer:"
`;

exports[`prompt/templates/chat question-answer flatten 1`] = `
"You are a research assistant in the "{{domain}}" domain. Based on the following rules and provided corpus denoted by ''', answer the question. 
Rules:
{{#each rules}}{{add @index 1}}. {{.}}
{{/each}}

Corpus: '''
{{context}}
'''

Question: {{question}}

Answer: "
`;

exports[`prompt/templates/chat question-answer messages 1`] = `
"You are a research assistant in the "DOMAIN" domain. Based on the following rules and provided corpus denoted by ''', answer the question. 
Rules:
1. only use knowledge from the provided corpus to answer the question
2. always be truthful, honest, unbiased, and unharmful
3. be concise, do not repeat the question or yourself in the answer

Corpus: '''
CORPUS
'''
Question: THE QUESTION

Answer:"
`;

exports[`prompt/templates/chat question-answer partial template overrides 1`] = `
"<WRAPPER>You are a research assistant in the "DOMAIN" domain. Based on the following rules and provided corpus denoted by ''', answer the question. 
Rules:
1. only use knowledge from the provided corpus to answer the question
2. always be truthful, honest, unbiased, and unharmful
3. be concise, do not repeat the question or yourself in the answer

Corpus: '''
CORPUS
'''
Question: THE QUESTION

Answer: </WRAPPER>"
`;

exports[`prompt/templates/chat question-answer wrappers 1`] = `
"<SEQUENCE>You are a research assistant in the "DOMAIN" domain. Based on the following rules and provided corpus denoted by ''', answer the question. 
Rules:
1. only use knowledge from the provided corpus to answer the question
2. always be truthful, honest, unbiased, and unharmful
3. be concise, do not repeat the question or yourself in the answer

Corpus: '''
CORPUS
'''
Question: THE QUESTION

Answer: </SEQUENCE>"
`;
